memstore/kafka_redolog_manager.go

```diff
@@ -0,0 +1,168 @@
+package memstore
+
+import (
+	"encoding/json"
+	"github.com/Shopify/sarama"
+	"github.com/uber/aresdb/utils"
+	"math"
+	"sync"
+)
+
+const maxBatchesPerFile = 5000
+
+type upsertBatchBundle struct {
+	batch       *UpsertBatch
+	kafkaOffset int64
+}
+
+// kafkaRedologManager is the implementation of RedologFileRotator
+type kafkaRedologManager struct {
+	sync.RWMutex
+
+	Topic     string `json:"topic"`
+	TableName string `json:"table"`
+	Shard     int    `json:"shard"`
+
+	MaxEventTimePerFile map[int64]uint32 `json:"maxEventTimePerFile"`
+
+	consumer          sarama.Consumer
+	partitionConsumer sarama.PartitionConsumer
+
+	commitFunc func(offset int64) error
+	done       chan struct{}
+	batches    chan upsertBatchBundle
+}
+
+// NewKafkaRedologManager creates kafka redolog manager
+func NewKafkaRedologManager(namespace, table string, shard int, consumer sarama.Consumer, commitFunc func(offset int64) error) KafkaRedologManager {
+	topic := utils.GetTopicFromTable(namespace, table)
+	return &kafkaRedologManager{
+		TableName:           table,
+		Shard:               shard,
+		Topic:               topic,
+		consumer:            consumer,
+		commitFunc:          commitFunc,
+		MaxEventTimePerFile: make(map[int64]uint32),
+		done:                make(chan struct{}),
+		batches:             make(chan upsertBatchBundle, 1),
+	}
+}
+
+// WriteUpsertBatch to kafka redolog manager is disabled
+func (k *kafkaRedologManager) WriteUpsertBatch(upsertBatch *UpsertBatch) (int64, uint32) {
+	panic("WriteUpsertBatch to kafka redolog manager is disabled")
+}
+
+func (k *kafkaRedologManager) UpdateMaxEventTime(eventTime uint32, fileID int64) {
+	k.Lock()
+	defer k.Unlock()
+	if _, ok := k.MaxEventTimePerFile[fileID]; ok && eventTime <= k.MaxEventTimePerFile[fileID] {
+		return
+	}
+	k.MaxEventTimePerFile[fileID] = eventTime
+}
+
+// NextUpsertBatch creates iterator function for getting the next upsert batch
+func (k *kafkaRedologManager) NextUpsertBatch() func() (*UpsertBatch, int64, uint32) {
+	return func() (*UpsertBatch, int64, uint32) {
+		for {
+			select {
+			case batchBundle := <-k.batches:
+				fileID := batchBundle.kafkaOffset / maxBatchesPerFile
+				fileOffset := batchBundle.kafkaOffset % maxBatchesPerFile
+				return batchBundle.batch, fileID, uint32(fileOffset)
+			case <-k.done:
+				// redolog manager closed
+				return nil, 0, 0
+			}
+		}
+	}
+}
+
+func (k *kafkaRedologManager) PurgeRedologFileAndData(eventTimeCutoff uint32, fileIDCheckpointed int64, batchOffset uint32) error {
```

Comment:
it doesn't purge data in this implementation, maybe add a comment, if renaming interface is tricky

Reply:
change to CheckpointRedolog